### PR TITLE
Sep state vars

### PR DESCRIPTION
**Pull request by RobertGemmaJr**
Separates `colorMaps`, `model`, `transferFunction`, etc. into their own state variables. This will make separating the app into different exports much easier but that's left for another PR

### CODE REVIEW

**Comment by RobertGemmaJr** [comment]
The reset button is moved up into `Controls.jsx` and now remounts the entire component when clicked. This resets the model's position in the scene

File: src/components/controls/Controls.jsx [hunk 1]
```diff
@@ -6,29 +6,35 @@ import OpacityControls from "./OpacityControls.jsx";
...
         <OpacityControls
-          state={state}
-          setState={setState}
-          initColorMap={initColorMap}
+          range={model.range}
+          initTransferFunction={initTransferFunction}
+          setTransferFunction={setTransferFunction}
         />
       )}
+      <Button onClick={reset}> Reset </Button>
```
Comment (mcmcgrath13) [inline src/components/controls/Controls.jsx:35]: does reset reset everything? or just the camera angle?

**Review by mcmcgrath13 — APPROVED** [review]
🙌

File: src/components/VolumeViewer.jsx [hunk 2]
```diff
@@ -16,74 +16,94 @@ import AframeScene from "./AframeScene.jsx";
...
-      z: [SLIDER_RANGE.min, SLIDER_RANGE.max],
-    },
-    transferFunction: getTransferFunction(),
-  });
-
-  // Update colorMap on prop change
+  // Changing a components key will remount the entire thing
+  // Because the model's position is handled internally by aframe we
+  // need to remount it to reset its position
+  const [remountKey, setRemountKey] = useState(Math.random());
```
Comment (broarr) [inline src/components/VolumeViewer.jsx:30]: That's an interesting way of forcing a rerender!

File: src/components/VolumeViewer.jsx [hunk 3]
```diff
@@ -16,74 +16,94 @@ import AframeScene from "./AframeScene.jsx";
...
+
+  // Control colorMap in state and update on prop change
+  const getColorMap = useCallback(() => {
+    return colorMapProp ?? DEFAULT_COLOR_MAP;
+  }, [colorMapProp]);
+  const [colorMap, setColorMap] = useState(getColorMap());
```
Comment (broarr) [inline src/components/VolumeViewer.jsx:36]: You did this in another PR. I'm not sure I understand why you're creating a color map callback and then storing that in state. Couldn't you store `const [colorMap, setColorMap] = useState(colorMapProp ?? DEFAULT_COLOR_MAP);`?

File: src/components/VolumeViewer.jsx [hunk 4]
```diff
@@ -16,74 +16,94 @@ import AframeScene from "./AframeScene.jsx";
...
   useEffect(() => {
-    setState((state) => ({
-      ...state,
-      colorMap: getColorMap(),
-    }));
-  }, [colorMap]);
-
-  // Update model on prop change
+    setColorMap(getColorMap());
```
Comment (broarr) [inline src/components/VolumeViewer.jsx:38]: Can this line be `setColorMap(colorMapProp ?? DEFAULT_COLOR_MAP)`? Do you have the callback function to DRY up this code?

**Review by broarr — COMMENTED** [review]
Some questions about react stuff, but this looks good otherwise

File: src/components/controls/Controls.jsx (hunk 1 above)
Comment (RobertGemmaJr) [reply src/components/controls/Controls.jsx:35]: Yup! The state reset is defined in `VolumeViewer.jsx`

File: src/components/VolumeViewer.jsx (hunk 2 above)
Comment (RobertGemmaJr) [reply src/components/VolumeViewer.jsx:30]: Stack overflow seems to agree it's the easiest way of doing so

File: src/components/VolumeViewer.jsx (hunk 3 above)
Comment (RobertGemmaJr) [reply src/components/VolumeViewer.jsx:36]: The logic ends up much more complicated down the line (plenty enough to warrant a separate function imo), it's really just skeleton code at this point

Comment (RobertGemmaJr) [reply src/components/VolumeViewer.jsx:36]: #40

File: src/components/VolumeViewer.jsx (hunk 4 above)
Comment (RobertGemmaJr) [reply src/components/VolumeViewer.jsx:38]: Yeah same as above - when things get more complicated it's a lot of logic to repeat

**Comment by RobertGemmaJr** [comment]
RE the `useEffect` and `useCallback` it still seems like this is the best way to handle it. If I do `const [model, setModel] = useState({ ...DEFAULT_MODEL, ...modelProp });` for example, it won't re-fire when `modelProp` changes. It needs to be a `setModel` and in order to do that without constantly forcing a re-render it needs to be inside a `useEffect`.

The only simplification I could do is remove the function (`getModel`) from the `useEffect` dependency array. This would let me eliminate the `usecallback` step (making it a basic arrow function) but ESLint shows me a warning when the function is missing from the dependency array.

I'm learning on the fly with all this so I very well might not be looking at the best solution but I don't know any way to change the props from another component.

**Comment by mcmcgrath13** [comment]
`useEffect` with the correct dependencies sounds pretty idiomatic to me.  You can put a comment on the line about the dependencies to tell eslint to ignore that error.  eslint is not all knowing 😉

**Comment by RobertGemmaJr** [comment]
Actually I think I found a better solution. Having the getter function inside the closure meant React was re-creating the function definition for each render. Pulling them outside of the closure means there's a single function definition, I just need to pass the prop in as a parameter since it's no longer in the same scope. Since the function will never change it's not actually a dependency for `useEffect` and, even better, the function will get called whenever the prop changes so I can use that return to update state. It's a similar solution as `toAframeString` in the other branch and is way cleaner.

I'm not sure what to call the return variable though. They're prepended with `init` right now but I'm not sure that's the clearest name

**Review by mcmcgrath13 — APPROVED** [review]
🥇

**Pull request merged**